students/gregdevore/lesson03/mailroom.py

```diff
@@ -0,0 +1,134 @@
+
+def add_donation(donor_name, donation_amount):
+    # Find donor in database
+    donor_names = get_donor_names()
+    donor_index = donor_names.index(donor_name)
+    # Get donation history and add latest donation
+    donations = donors[donor_index][1]
+    donations.append(donation_amount)
+    # Note: Even though each donor's data is stored as a tuple, which is immutable,
+    #       their donation history is stored as a list, which is mutable. The tuple only
+    #       contains a reference to the list, so the list can be changed even
+    #       though it resides inside of a tuple. Hence, the above append statement works.
+    return
+
+def generate_email(donor_name, donation_amount):
+    # Create formatted email that can be copied & pasted
+    email = ('\n'.join(['','Dear {donor_name},','',
+    'Thank you for your generous donation of ${donation_amount:.2f}.',
+    'Your contribution will help new arrivals receive the highest quality care possible.',
+    'Please know that your donation makes a world of difference!',
+    '','Sincerely,','The Good Place Team'])).format(donor_name=donor_name,donation_amount=donation_amount)
+    print(email)
+    return
+
+def write_thank_you():
+    # Add donation for new or existing donor and compose 'Thank You' message
+    # Get current list of donors
+    donor_names = get_donor_names()
+    while True:
+        print()
+        name = input('Enter donor name: ')
+        if name == 'list': # List donors
+            print('\nCurrent list of donors:\n')
+            print('\n'.join(donor_names))
+            print()
+        elif name == 'quit': # Return to main prompt
+            return
+        else:
+            new_donor = False
+            if name not in donor_names: # New donor, flag as new and add to database
+                new_donor = True
+                donors.append((name, []))
+            # Prompt for donation amount
+            amount = input('Enter donation amount in dollars: ')
+            # If the user wants to bail mid-entry, remove the donor that was just
+            # added (if they were new) and return to main prompt
+            if amount == 'quit':
+                if new_donor:
+                    donors.pop()
+                return
+            else: # Otherwise, convert donation amount to float
+                amount = float(amount)
+            # Add donation to database
+            add_donation(name, amount)
+            # Generate email
+            generate_email(name, amount)
+            # Return to main program
+            return
+
+def donor_key(donor):
+    # Donor is a tuple of the form (name, total donation, number of donations, average donation)
+    # Sort by total donation
+    return donor[1]
+
+def generate_report_data():
+    donor_names = get_donor_names()
+    total_donation = [sum(data[1]) for data in donors ]
+    num_donation = [len(data[1]) for data in donors]
+    avg_donation = [sum(data[1])/len(data[1]) for data in donors]
```

Comment:
I would certainly recommend to use a for loop here since you're looping over the same data structure 3 separate time which is wasteful.